lib/saluki-io/src/deser/codec/dogstatsd/mod.rs

```diff
@@ -457,6 +472,82 @@ fn parse_dogstatsd_event<'a>(input: &'a [u8], config: &DogstatsdCodecConfigurati
     Ok((remaining, eventd))
 }
 
+fn parse_dogstatsd_service_check<'a>(
+    input: &'a [u8], config: &DogstatsdCodecConfiguration,
+) -> IResult<&'a [u8], ServiceCheck> {
+    let (remaining, (name, raw_check_status)) = preceded(
+        tag(message::SERVICE_CHECK_PREFIX),
+        separated_pair(ascii_alphanum_and_seps, tag(b"|"), parse_u8),
+    )(input)?;
+
+    let check_status = saluki_event::service_check::CheckStatus::try_from(raw_check_status)
+        .map_err(|_| nom::Err::Error(Error::new(input, ErrorKind::Verify)))?;
+
+    let mut maybe_timestamp = None;
+    let mut maybe_hostname = None;
+    let mut maybe_tags = None;
+    let mut maybe_message = None;
+    let mut seen_message = false;
+
+    let remaining = if !remaining.is_empty() {
+        let (mut remaining, _) = tag("|")(remaining)?;
+        while let Some((chunk, tail)) = split_at_delimiter(remaining, b'|') {
+            if chunk.len() < 2 {
+                break;
+            }
+
+            // Message field must be positioned last among the metadata fields but it was already seen
+            if seen_message {
```

Comment:
break if we already saw `m:` but are still parsing optional fields since `m:` [should be positioned last](https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/?tab=servicechecks). should we error instead?

Reply:
Conceptually, we should error, yes... but I think we should put the error in the catch-all match arm (`_ => {...}`) and then actually move the message parsing to outside of the loop, to make it a little clearer that we _always_ expect the message to 1) be present and 2) come as the very last field.

Reply:
> we always expect the message to 1) be present and 2) come as the very last field.

message is an optional field though.